### PR TITLE
Use pyfakefs and cleanup log loading

### PR DESCRIPTION
**Pull request by samtygier-stfc**
### Issue
Progresses #1219

### Description

This improves some of the tests around file finding and loading by using pyfakefs. I have made the tests a bit more rigorous by having reducing mocking and having them act on the fake filesystem. This exposed some issues around how paths are handled on windows (auto finding log files probably never worked). Switching some path handling to use pathlib required a cascade of small changes through a lot of files.

### Testing & Acceptance Criteria 

Failures are expected on the docker tests as the docker images wont have pyfakefs in yet. Other tests should pass.

From the Load Dataset dialog, select the flower dataset sample. Check that the log files get correctly detected:
image
When loading check that there are no
'Could not find a log file for' messages in the terminal.

### Documentation

Not needed yet

### CODE REVIEW

File: mantidimaging/core/io/loader/test/loader_test.py
Comment (JackEAllen) [inline mantidimaging/core/io/loader/test/loader_test.py:51]: NITPICK: can we use a more descriptive variable name than `ii` here

File: mantidimaging/core/io/loader/test/loader_test.py
Comment (JackEAllen) [inline mantidimaging/core/io/loader/test/loader_test.py:71]: Minor: Could we use parameterized to shorten this tests length?

File: mantidimaging/core/io/test/utility_test.py
Comment (JackEAllen) [inline mantidimaging/core/io/test/utility_test.py:14]: As this may be forgotten about in the future, we should probably include the removal of this as a requirement/acceptance criteria for the final issue related to this work and maybe include it as a code snippet in our documentation somewhere for developers.

File: mantidimaging/core/io/utility.py
Comment (JackEAllen) [inline mantidimaging/core/io/utility.py:169]: NITPICK: Can we add a doctring to this function

File: mantidimaging/core/io/utility.py
Comment (JackEAllen) [inline mantidimaging/core/io/utility.py:170]: NITPICK: Could we make this variable name more descriptive maybe `filename_group`

File: mantidimaging/gui/windows/image_load_dialog/field.py
Comment (JackEAllen) [inline mantidimaging/gui/windows/image_load_dialog/field.py:38]: NITPICK: If we are touching this function anyway, could we add a docstring

File: mantidimaging/gui/windows/image_load_dialog/field.py
Comment (JackEAllen) [inline mantidimaging/gui/windows/image_load_dialog/field.py:187]: NITPICK: Could we add a docstring here while we are touching this function

File: mantidimaging/gui/windows/image_load_dialog/field.py
Comment (JackEAllen) [inline mantidimaging/gui/windows/image_load_dialog/field.py:200]: NITPICK: Could we add a docstring here while we are touching this function

File: mantidimaging/gui/windows/image_load_dialog/presenter.py
Comment (JackEAllen) [inline mantidimaging/gui/windows/image_load_dialog/presenter.py:154]: NITPICK: Could we add a docstring here while we are touching this function

File: mantidimaging/gui/windows/image_load_dialog/presenter.py
Comment (JackEAllen) [inline mantidimaging/gui/windows/image_load_dialog/presenter.py:202]: NITPICK: Could we add a docstring here while we are touching this function

File: mantidimaging/gui/windows/image_load_dialog/test/presenter_test.py
Comment (JackEAllen) [inline mantidimaging/gui/windows/image_load_dialog/test/presenter_test.py:39]: Not for now as it would be scope creep, but I think this test could be broken up into more compartmentalised tests making it easier to understand failures when and if they occur

**Review by JackEAllen — COMMENTED** [review]
From a first visual pass through of your changes, they look good! 👍  
I've left a few nitpicks and minor suggestions, but nothing too important. I've also highlighted a few areas where we may want to consider writing some new starter or tech dept related issues to clean up a few things in the future as they are outside the scope of this PR.
I'll give this second visual pass through once i've tested your changes work functionally 🙂

File: mantidimaging/core/io/loader/loader.py
Comment (samtygier-stfc) [reply mantidimaging/core/io/loader/loader.py:178]: Yes, that's one of the big targets of the file handling refactoring/cleanup.

File: mantidimaging/core/io/loader/test/loader_test.py
Comment (samtygier-stfc) [reply mantidimaging/core/io/loader/test/loader_test.py:17]: I'll remove these for now, and have opened #1640

File: mantidimaging/core/io/loader/test/loader_test.py
Comment (samtygier-stfc) [reply mantidimaging/core/io/loader/test/loader_test.py:71]: Possibly, but it should get nicer once create_loading_parameters_for_file_path is cleaned up.

File: mantidimaging/core/io/utility.py
Comment (samtygier-stfc) [reply mantidimaging/core/io/utility.py:169]: I think for a short function like this a docstring doesn't add much that isn't in the signature.

File: mantidimaging/gui/windows/image_load_dialog/test/presenter_test.py
Comment (samtygier-stfc) [reply mantidimaging/gui/windows/image_load_dialog/test/presenter_test.py:39]: I'm hoping this will also get better with further cleanups

File: mantidimaging/core/io/loader/test/loader_test.py
Comment (JackEAllen) [reply mantidimaging/core/io/loader/test/loader_test.py:69]: Happy to hold off to see if it gets nicer as you say once `create_loading_parameters_for_file_path` has been cleaned up

File: mantidimaging/core/io/utility.py
Comment (JackEAllen) [reply mantidimaging/core/io/utility.py:169]: I'm happy to accept this. More my own personal preference I guess, but my general rule is to have a docstring for all methods small or large as it's speeds up code navigation for me being able to read a formal description over relying on the name truly being self explanatory to avoid any misinterpretations from method naming.

**Review by JackEAllen — APPROVED** [review]
I can confirm all tests pass and i'm happy with the changes made so this should be fine to merge in provided the rerun of the Windows build passes 👍

File: mantidimaging/core/io/test/utility_test.py
Comment (JackEAllen) [reply mantidimaging/core/io/test/utility_test.py:26]: Thanks ❤️

**Review by JackEAllen — APPROVED** [review]

**Pull request merged**
